scene/gui/text_edit.cpp

```diff
@@ -3897,7 +3925,90 @@ void TextEdit::delete_selection() {
 	update();
 }
 
-/* line wrapping. */
+/* Line and character position. */
+Vector2i TextEdit::_get_line_drawing_wrap_offsets(int p_line, int p_column) const {
+	ERR_FAIL_INDEX_V(p_line, text.size(), Vector2i(-1, -1));
+	ERR_FAIL_COND_V(p_column < 0, Vector2i(-1, -1));
+	ERR_FAIL_COND_V(p_column > text[p_line].length(), Vector2i(-1, -1));
+
+	if (line_drawing_cache.size() == 0 || !line_drawing_cache.has(p_line)) {
+		// Line not in the cache, which means it's outside of the viewing area.
+		return Vector2i(-1, -1);
+	}
+
+	LineDrawingCache cache_entry = line_drawing_cache[p_line];
+
+	/* Loop through wraps in the line text until we get to the column. */
+	int wrap_index = 0;
+	int wrap_column = p_column;
+	if (is_line_wrapped(p_line)) {
+		Vector<String> lines = get_line_wrapped_text(p_line);
+		for (int i = 0; i < lines.size(); i++) {
+			wrap_index = i;
+			String s = lines[wrap_index];
+
+			if (wrap_column < s.length()) {
+				break;
+			}
+			wrap_column -= s.length();
+		}
+	}
+
+	if (wrap_index >= cache_entry.first_visible_char.size()) {
+		// Line seems to be wrapped beyond the viewable area.
+		return Vector2i(-1, -1);
+	}
+
+	int first_visible_char = cache_entry.first_visible_char[wrap_index];
+	int last_visible_char = cache_entry.last_visible_char[wrap_index];
+	if (wrap_column < first_visible_char || wrap_column > last_visible_char) {
+		// Character is outside of the viewing area.
+		return Vector2i(-1, -1);
+	}
+
+	return Vector2i(wrap_column, wrap_index);
+}
+
+Point2i TextEdit::get_pos_at_line_column(int p_line, int p_column) const {
+	Vector2i wrap_offsets = _get_line_drawing_wrap_offsets(p_line, p_column);
+	if (wrap_offsets.x == -1 || wrap_offsets.y == -1) {
+		return Point2i(-1, -1);
+	}
+
+	Point2i pos;
+	LineDrawingCache cache_entry = line_drawing_cache[p_line];
+	pos.y = cache_entry.y_offset + get_line_height() * (1 + wrap_offsets.y);
+	pos.x = get_total_gutter_width() + style_normal->get_margin(SIDE_LEFT) - get_h_scroll();
+
+	RID text_rid = text.get_line_data(p_line)->get_line_rid(wrap_offsets.y);
+	Vector2 col_bounds = TS->shaped_text_get_grapheme_bounds(text_rid, p_column);
+	pos.x += col_bounds.x;
+
+	return pos;
+}
```

Comment:
> it's probably better to just get bounds and check if it's in the range, instead.

Something like:

```suggestion
Point2i TextEdit::get_pos_at_line_column(int p_line, int p_column) const {
	Point2i pos;
	LineDrawingCache cache_entry = line_drawing_cache[p_line];
	pos.y = cache_entry.y_offset + get_line_height() * (1 + wrap_offsets.y);
	pos.x = get_total_gutter_width() + style_normal->get_margin(SIDE_LEFT) - get_h_scroll();

	RID text_rid = text.get_line_data(p_line)->get_line_rid(wrap_offsets.y);
	Vector2 col_bounds = TS->shaped_text_get_grapheme_bounds(text_rid, p_column);
	pos.x += col_bounds.x;
	if (pos.x < 0 || pos.x > {MAX_WIDTH}) {
		return Point2i(-1, -1);
	}
	return pos;
}
```

Reply:
This indeed can be simplified as I no longer need `_get_line_drawing_wrap_offsets` with the new `TextServer` method.